nengo_spa/compiler/tests/test_parsing.py

```diff
@@ -0,0 +1,143 @@
+        parsing.Terminal(type=tk.NAME, string='bar')),
+    parsing.Terminal(type=tk.NAME, string='foo') |
+    parsing.Terminal(type=tk.NAME, string='bar')])
+def test_either(rule):
+    tokens = peekable_tokens('foo bar')
+    assert rule.accept(tokens)
+    assert rule.read(tokens) == [
+        MatchToken(type=tk.NAME, string='foo')]
+
+    assert rule.accept(tokens)
+    assert rule.read(tokens) == [
+        MatchToken(type=tk.NAME, string='bar')]
+
+    assert not rule.accept(tokens)
+    with pytest.raises(parsing.RuleMismatchError):
+        rule.read(tokens)
+
+
+def test_either_chaining_gathers_rules_in_single_object():
+    rules = [parsing.Terminal() for i in range(3)]
+    either = rules[0] | rules[1] | rules[2]
+    assert all(r1 is r2 for r1, r2 in zip(either.rules, rules))
+
+
+@pytest.mark.parametrize('rule', [
+    parsing.Chain(
+        parsing.Terminal(type=tk.NAME, string='foo'),
+        parsing.Terminal(type=tk.NAME, string='bar')),
+    parsing.Terminal(type=tk.NAME, string='foo') +
+    parsing.Terminal(type=tk.NAME, string='bar')])
+def test_chain(rule):
+    tokens = peekable_tokens('foo bar')
+    assert rule.accept(tokens)
+    assert rule.read(tokens) == [
+        MatchToken(type=tk.NAME, string='foo'),
+        MatchToken(type=tk.NAME, string='bar')]
+
+    assert not rule.accept(tokens)
+    with pytest.raises(parsing.RuleMismatchError):
+        rule.read(tokens)
+
+
+def test_chaining_gathers_rules_in_single_object():
+    rules = [parsing.Terminal() for i in range(3)]
+    either = rules[0] + rules[1] + rules[2]
+    assert all(r1 is r2 for r1, r2 in zip(either.rules, rules))
+
+
+def test_maybe():
+    maybe = parsing.Maybe(parsing.Terminal(type=tk.NAME, string='foobar'))
+    tokens = peekable_tokens('foobar')
+    assert maybe.accept(tokens)
+    assert maybe.read(tokens) == [MatchToken(type=tk.NAME, string='foobar')]
+
+    assert maybe.accept(tokens)
+    assert maybe.read(tokens) == []
+
+
+@pytest.mark.parametrize('code', ['foobar', 'foo bar'])
+def test_at_least_one(code):
+    tokens = peekable_tokens(code)
+
+    at_least_one = parsing.AtLeastOne(parsing.Terminal(type=tk.NAME))
+    assert at_least_one.accept(tokens)
+    parse = at_least_one.read(tokens)
+    expected = code.split(' ')
+    assert [x[1] for x in parse] == expected
+
+    assert not at_least_one.accept(tokens)
+    with pytest.raises(parsing.RuleMismatchError):
+        at_least_one.read(tokens)
+
+
+@pytest.mark.parametrize('code', ['', 'foobar', 'foo bar'])
+def test_any_number(code):
+    tokens = peekable_tokens(code)
+
+    any_number = parsing.AnyNumber(parsing.Terminal(type=tk.NAME))
+    assert any_number.accept(tokens)
+    parse = any_number.read(tokens)
+    expected = code.split(' ')
+    if expected == ['']:
+        expected = []
+    assert [x[1] for x in parse] == expected
+
+
+def test_group():
+    tokens = peekable_tokens('foo bar')
+    grammar = parsing.Group('grammar', parsing.Group(
```

Comment:
Is `grammar` and `word` a significant identifier here or could you also just say `parent` and `child`?